### PR TITLE
[dask] [python-package] include support for column array as label

### PR DESCRIPTION
**Pull request by jmoralez**
This solves #3909 by allowing to have an `(n_samples, 1)` array as label when `data` is a `pandas.DataFrame`.

### CODE REVIEW

**Comment by jmoralez** [comment]
Hi @jameslamb. Can you give me your thoughts on these changes?

**Comment by jameslamb** [comment]
Thanks so much for looking into this!

>  Can you give me your thoughts on these changes?

Can you please add tests (one in `test_sklearn.py`, one in `test_dask.py`) testing that training is successful and correct when `data` is a `pandas` DataFrame and `label` is a numpy column array? Once you've written tests like that proving that this PR fixes the issue described in #3909, I'd be happy to give a more thorough review on the approach.

File: tests/python_package_test/test_basic.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 import lightgbm as lgb
 import numpy as np
+import pandas as pd
```
Comment (StrikerRUS) [inline tests/python_package_test/test_basic.py:6]: @jmoralez Please make `pandas` optional for basic tests like you did in sklearn tests:
```
pd = pytest.importorskip("pandas")
```

File: tests/python_package_test/test_dask.py [hunk 2]
```diff
@@ -1108,3 +1108,26 @@ def test_dask_methods_and_sklearn_equivalents_have_similar_signatures(methods):
     for param in dask_spec.args:
         error_msg = f"param '{param}' has different default values in the methods"
         assert dask_params[param].default == sklearn_params[param].default, error_msg
+
+
+def test_training_succeeds_when_data_is_dataframe_and_label_is_column_array(
+    client,
+    listen_port
+):
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:1116]: thanks for this! I think test could be a bit stronger. Instead of just smoke-testing that training runs without error, could you make this test check that `label` vectors of shape `[n_samples, 1]` and `[n_samples]` result in identical models?

I think that could look something like this

```python
    params = {
        'n_estimators': 1,
        'num_leaves': 3,
        'local_listen_port': listen_port,
        'time_out': 5
    }
    _, _, _, dX, dy, dw = _create_data(
        objective='regression',
        output='dataframe'
    )
    dy = dy.to_dask_array(lengths=True)
    dy_col_array = dy.reshape(-1, 1)
    assert len(dy_col_array.shape) == 2 and dy_col_array.shape[1] == 1

    model1 = lgb.DaskLGBMRegressor(**params)
    model1.fit(dX, dy, sample_weight=dw)
    p1 = model1.predict(dX).compute()
    assert model1.fitted_
    
    dy_vector = dy.reshape(-1)
    assert len(dy_col_array.shape) == 1
    model2 = lgb.DaskLGBMRegressor(**params)
    model2.fit(dX, dy, sample_weight=dw)
    p2 = model2.predict(dX).compute()
    assert model2.fitted_
    
    assert_eq(p1, p2)
```

Notice in the above that I also changed the check from `assert model.booster_` to `assert model.fitted_`. I think that's a more direct check of the thing we're trying to check ("is the model fitted").

File: tests/python_package_test/test_dask.py [hunk 3]
```diff
@@ -1108,3 +1108,26 @@ def test_dask_methods_and_sklearn_equivalents_have_similar_signatures(methods):
     for param in dask_spec.args:
         error_msg = f"param '{param}' has different default values in the methods"
         assert dask_params[param].default == sklearn_params[param].default, error_msg
+
+
+def test_training_succeeds_when_data_is_dataframe_and_label_is_column_array(
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:1113]: can you please run this test for all three of `DaskLGBMClassifier`, `DaskLGBMRanker`, and `DaskLGBMRegressor`? Similar to https://github.com/microsoft/LightGBM/blob/84c4b7504a49eed63fa819a92db7d0b18292f3f4/tests/python_package_test/test_dask.py#L966

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks! The fix looks good to me. I left some suggestions for making the tests a bit stronger.

File: tests/python_package_test/test_basic.py (hunk 1 above)
Comment (jmoralez) [reply tests/python_package_test/test_basic.py:6]: Since I had some `pd.Series` in the parametrize for `y` I changed it to this:
```python
    ('y', 'is_series'),
    [
        (np.random.rand(10), False),
        (np.random.rand(10, 1), False),
        ([1] * 10, False),
        (np.random.rand(10), True),
        (['a', 'b'], True),
        ([[1], [2]], False)
    ])
```
and then I do this inside the function
```python
    if is_series:
        pd = pytest.importorskip('pandas')
        y = pd.Series(y)
```
Is that the best way to do it?

File: tests/python_package_test/test_dask.py (hunk 3 above)
Comment (jmoralez) [reply tests/python_package_test/test_dask.py:1113]: This only seems to impact the regressor, the ranker and the classifier go through scikit-learns check_X_y

Comment (jameslamb) [reply tests/python_package_test/test_dask.py:1113]: testing all three will prevent us from introducing a change in the future that makes that not true

Comment (jmoralez) [reply tests/python_package_test/test_dask.py:1113]: Yeah that sounds reasonable. I'll include it in test_sklearn as well

Comment (jmoralez) [reply tests/python_package_test/test_dask.py:1113]: Is there a way to make the model deterministic? I'm calling repeatedly `model_1d.fit(dX, dy, sample_weight=dw).predict(dX).compute()` and the results differ from time to time. I've set `random_state` in the parameters.

Comment (jameslamb) [reply tests/python_package_test/test_dask.py:1113]: you can try adding "deterministic: true" to the parameters: https://lightgbm.readthedocs.io/en/latest/Parameters.html#deterministic

Comment (jmoralez) [reply tests/python_package_test/test_dask.py:1115]: I had to switch to single worker training for the test to pass everytime

Comment (jameslamb) [reply tests/python_package_test/test_dask.py:1115]: ok, I'd prefer removing the `assert_eq()` then and going back to multi-worker training, sorry. It can be difficult to guarantee consistent results when doing distributed training on such tiny data sizes. I had hoped it would work but if it doesn't, we can go back to just training one model and checking `.fitted_`. Sorry for sending you down a road to a dead end.

Comment (jmoralez) [reply tests/python_package_test/test_dask.py:1115]: Sorry hadn't seen this. Even though the training is only happening in one worker the idea is testing that both the dask and scikit-learn models can train with a column array and get the same results, isn't it?

Comment (jameslamb) [reply tests/python_package_test/test_dask.py:1115]: For the Dask interface, having data in a single partition is basically just falling back to training with the sklearn estimators.

The test you've added in `test_sklearn.py` that ends in `assert np.allclose(preds_1d, preds_2d)` is already checking that, for non-distributed training with the sklearn interface, the model output is the same when a label has shape `[n_samples, 1]` or `[n_samples]`.

Given that, I'd rather have the Dask test be a smoke test that multi-worker distributed training works when data has a column-array shape.

File: tests/python_package_test/test_dask.py [hunk 4]
```diff
@@ -196,6 +196,15 @@ def _unpickle(filepath, serializer):
         raise ValueError(f'Unrecognized serializer type: {serializer}')
 
 
+def collection_to_single_partition(collection):
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply tests/python_package_test/test_dask.py:199]: you can ignore this comment. Please just put this method back, based on https://github.com/microsoft/LightGBM/pull/3943/files#r575617861

File: tests/python_package_test/test_basic.py (hunk 1 above)
Comment (StrikerRUS) [reply tests/python_package_test/test_basic.py]: Maybe it will be easier to just mark the whole test for skipping if `pandas` is not installed?
```
@pytest.mark.skipif(not lgb.compat.PANDAS_INSTALLED, reason='pandas is not installed')
```
like here
https://github.com/microsoft/LightGBM/blob/50e061f396d7c1b704e79371dea0ddf07680ba75/tests/python_package_test/test_plotting.py#L132

Yeah, it will skip the whole test if pandas is not installed but it looks much easier to read and maintain than one more parametrization argument.

Comment (jameslamb) [reply tests/python_package_test/test_basic.py]: I agree with @StrikerRUS , I think that's simpler than having this flag and it's ok. We install `pandas` in CI so the tests won't be skipped there.

Comment (jmoralez) [reply tests/python_package_test/test_basic.py]: But how would I define the series within the parametrize arguments? I can maybe make a second test that runs for series only and skip that one if pandas isn't installed. Or maybe just skip the `test_list_to_1d_numpy` altogether and just test the training with the column array?

Comment (StrikerRUS) [reply tests/python_package_test/test_basic.py]: I believe you could use `pd_Series` class from our `compat.py` module.
https://github.com/microsoft/LightGBM/blob/4ae59494ab9091a8405a61ae7b13bdedbdbed93e/python-package/lightgbm/compat.py#L4-L17

Will this work?
```py
@pytest.mark.skipif(not PANDAS_INSTALLED, reason='pandas is not installed') 
@pytest.mark.parametrize(
    'y',
    [
        (np.random.rand(10)),
        (np.random.rand(10, 1)),
        ([1] * 10),
        (pd_Series(np.random.rand(10))),
        (pd_Series(['a', 'b'])),
        ([[1], [2]])
    ])
@pytest.mark.parametrize('dtype', [np.float32, np.float64])
def test_list_to_1d_numpy(y, dtype):
```

Comment (jmoralez) [reply tests/python_package_test/test_basic.py]: I hadn't seen that, that definitely works, thank you!

**Review by jameslamb — APPROVED** [review]
changes look good to me, thanks for picking this up!

**Comment by jmoralez** [comment]
Thank you for your help, James! I just realized I wasn't fitting to the column array in the test for dask so I fixed that.

**Comment by StrikerRUS** [comment]
@jmoralez 
Sorry for the inconvenience. We just merged #3947. Could you please resolve conflicts?

**Comment by jameslamb** [comment]
@jmoralez sorry again, but can you please merge `master`  into this? We just fixed some additional CI issues 😬

File: python-package/lightgbm/basic.py [hunk 5]
```diff
@@ -126,6 +126,21 @@ def is_numpy_1d_array(data):
...
+def is_numpy_column_array(data):
+    """Check whether data is a column numpy array."""
+    if not isinstance(data, np.ndarray):
+        return False
+    shape = data.shape
+    return len(shape) == 2 and shape[1] == 1
+
+
+def numpy_1d_array_to_dtype(array, dtype):
+    """Convert 1d array to dtype."""
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py:138]:
```suggestion
def numpy_1d_array_cast_to_dtype(array, dtype):
    """Cast 1d array to given dtype."""
```

File: python-package/lightgbm/basic.py [hunk 6]
```diff
@@ -134,10 +149,11 @@ def is_1d_list(data):
...
     if is_numpy_1d_array(data):
-        if data.dtype == dtype:
-            return data
-        else:
-            return data.astype(dtype=dtype, copy=False)
+        return numpy_1d_array_to_dtype(data, dtype)
+    elif is_numpy_column_array(data):
+        _log_warning('Converting column-vector to 1d array')
+        array = data.ravel()
+        return numpy_1d_array_to_dtype(array, dtype)
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py:156]:
```suggestion
        return numpy_1d_array_cast_to_dtype(array, dtype)
```

File: tests/python_package_test/test_sklearn.py [hunk 7]
```diff
@@ -1192,3 +1192,36 @@ def test_parameters_default_constructible(estimator):
...
+        if task == 'ranking':
+            model_1d = model_factory(**params).fit(X, y, group=g)
+            model_2d = model_factory(**params).fit(X, y_col_array, group=g)
+        else:
+            model_1d = model_factory(**params).fit(X, y)
+            model_2d = model_factory(**params).fit(X, y_col_array)
+
+    preds_1d = model_1d.predict(X)
+    preds_2d = model_2d.predict(X)
+    assert np.allclose(preds_1d, preds_2d)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:1227]: We prefer `numpy.testing` module for asserting arrays. Also, I think it is important to test for exact equal because models should be identical given the difference in only shape of `y`.
```suggestion
    numpy.testing.assert_array_equal(preds_1d, preds_2d)
```

File: tests/python_package_test/test_dask.py [hunk 8]
```diff
@@ -1084,6 +1084,45 @@ def test_dask_methods_and_sklearn_equivalents_have_similar_signatures(methods):
         assert dask_params[param].default == sklearn_params[param].default, error_msg
 
 
+@pytest.mark.parametrize('task', tasks)
+def test_training_succeeds_when_data_is_dataframe_and_label_is_column_array(
+    task,
+    client,
+    listen_port
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:1091]: @jameslamb Should `listen_port` be removed here given that #3994 is merged now?

File: python-package/lightgbm/basic.py (hunk 5 above)
Comment (jmoralez) [reply python-package/lightgbm/basic.py:138]: How about `cast_numpy_1d_array_to_dtype`?

Comment (StrikerRUS) [reply python-package/lightgbm/basic.py:138]: Sounds good!

File: tests/python_package_test/test_dask.py (hunk 8 above)
Comment (jameslamb) [reply tests/python_package_test/test_dask.py:1091]: yes it should

**Review by StrikerRUS — APPROVED** [review]
Nice job! Thank you very much!

**Comment by jmoralez** [comment]
Thank you for your comments!
